src/Interpreters/DDLWorker.cpp

```diff
@@ -1146,7 +1157,12 @@ void DDLWorker::runMainThread()
         catch (const Coordination::Exception & e)
         {
             if (!Coordination::isHardwareError(e.code))
-                throw;  /// A logical error.
+            {
+                /// A logical error.
+                LOG_ERROR(log, "ZooKeeper error: {}. Failed to start DDLWorker.",getCurrentExceptionMessage(true));
+                reset_state(false);
+                assert(false);  /// Catch such failures in tests with debug build
```

Comment:
@tavplubix Maybe better to throw `LOGICAL_ERROR`?

Reply:
The purpose of these changes is to restart DDLWorker thread in case of any unexpected errors. If we throw any exceptions here then server continue works, but DDLWorker doesn't, so all ON CLUSTER queries hang until server is restarted manually.

Reply:
But `LOGICAL_ERROR` is the same as `assert` - https://github.com/ClickHouse/ClickHouse/blob/8fdaa43b0f9e0dacd83fb3c9dabea67f985c6e79/src/Common/Exception.cpp#L43-L49 (and not only in debug build but under santiziers it will also abort)

Reply:
> But LOGICAL_ERROR is the same as assert

No, in release builds `assert` does nothing, but `LOGICAL_ERROR` works as usual exception. If we throw any exception here then user have to manually restart ClickHouse to make DDLWorker work.